jupyterlab_git/handlers.py

```diff
@@ -436,32 +445,76 @@ async def post(self):
+            remote?: string # Remote to push to; i.e. <remote_name> or <remote_name>/<branch>
+        }
         """
         data = self.get_json_body()
         current_path = data["current_path"]
+        known_remote = data.get("remote")
 
         current_local_branch = await self.git.get_current_branch(current_path)
-        upstream = await self.git.get_upstream_branch(
+        
+        set_upstream = False
+        current_upstream_branch = await self.git.get_upstream_branch(
             current_path, current_local_branch
         )
 
-        if upstream['code'] == 0:
-            branch = ":".join(["HEAD", upstream['remote_branch']])
+        if known_remote is not None:
+            set_upstream = current_upstream_branch['code'] != 0
+
+            remote_name, _, remote_branch = known_remote.partition("/")
+            
+            current_upstream_branch = {
+                "code": 0,
+                "remote_branch": remote_branch or current_local_branch,
+                "remote_short_name": remote_name
+            }
+
+        if current_upstream_branch['code'] == 0:
+            branch = ":".join(["HEAD", current_upstream_branch['remote_branch']])
             response = await self.git.push(
-                upstream['remote_short_name'], branch, current_path, data.get("auth", None)
+                current_upstream_branch['remote_short_name'], branch, current_path, data.get("auth", None), set_upstream
             )
 
         else:
-            if ("no upstream configured for branch" in upstream['message'].lower()
-                 or 'unknown revision or path' in upstream['message'].lower()):
+            # Allow users to specify upstream through their configuration
+            # https://git-scm.com/docs/git-config#Documentation/git-config.txt-pushdefault
+            # Or use the remote defined if only one remote exists
+            config = await self.git.config(current_path)
+            config_options = config["options"]
+            list_remotes = await self.git.remote_show(current_path)
+            remotes = list_remotes.get("remotes", list())
+            push_default = config_options.get('remote.pushdefault')
+
+            default_remote = None
+            if push_default is not None and push_default in remotes:
+                default_remote = push_default
+            elif len(remotes) == 1:
+                default_remote = remotes[0]
+
+            if default_remote is not None:
+                response = await self.git.push(
+                    default_remote, 
+                    current_local_branch, 
+                    current_path, 
+                    data.get("auth", None), 
+                    set_upstream=True,
+                )
+            else:
                 response = {
                     "code": 128,
                     "message": "fatal: The current branch {} has no upstream branch.".format(
                         current_local_branch
```

Comment:
Maybe we should modify this error message to explain to the user why the guessing of the push target failed. Something along the lines of this
```suggestion
                response = {
                    "code": 128,
                    "message": "The current branch {} has no upstream branch. You have {} remotes so a remote cannot be chosen automatically without having set a pushdefault.".format(
                        current_local_branch,
                        len(remotes)
```

Reply:
That error message will be used also for a repository without remote.

So I will not change the error message. But prompt the user to choose from the remote list if it exists. Hence your use case will be covered. What do you think?

Reply:
> But prompt the user to choose from the remote list if it exist

That is the best solution by far. My suggestion was to make it more clear if there were to be no further frontend changes.

Reply:
Will having the option to prompt the user to choose from a remote list run into the same issues as the Git Credentials w.r.t the new modal?